src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -187,8 +187,8 @@ class IndexAllocation implements NativeMemoryAllocation {
 
         protected void closeInternal() {
             Runnable onClose = () -> {
+                writeLock();
                 try {
-                    writeLock();
```

Comment:
`writeLock` must out of try-finally to make sure there's a acquire-and-release.

Reply:
Why is this?

Reply:
If the method raised an exception (I know the current implementation doesn't), we should not cal `unlock`. 
`unlock` in the finally statement should be called once after we obtained a lock, which is why we need to put `writeLock` before try-finally.